zenoh/src/net/routing/interceptor/authorization.rs

```diff
@@ -0,0 +1,250 @@
+                } else {
+                    let policy_information = self.policy_information_point(rules)?;
+                    let subject_map = policy_information.subject_map;
+                    let mut main_policy: PolicyMap = PolicyMap::default();
+
+                    for rule in policy_information.policy_rules {
+                        if let Some(index) = subject_map.get(&rule.subject) {
+                            if let Some(single_policy) = main_policy.get_mut(index) {
+                                single_policy
+                                    .flow_mut(rule.flow)
+                                    .action_mut(rule.action)
+                                    .permission_mut(rule.permission)
+                                    .insert(keyexpr::new(&rule.key_expr)?, true);
+                            } else {
+                                let mut pfs = PolicyForSubject::default();
+                                pfs.flow_mut(rule.flow)
+                                    .action_mut(rule.action)
+                                    .permission_mut(rule.permission)
+                                    .insert(keyexpr::new(&rule.key_expr)?, true);
+                                main_policy.insert(*index, pfs);
+                            }
+                        };
+                    }
+                    self.policy_map = main_policy;
+                    self.subject_map = subject_map;
+                }
+            } else {
+                log::warn!("[ACCESS LOG]: No ACL rules have been specified!!!");
+            }
+        }
+        Ok(())
+    }
+
+    /*
+       converts the sets of rules from config format into individual rules for each subject, key-expr, action, permission
+    */
+    pub fn policy_information_point(
+        &self,
+        config_rule_set: &Vec<AclConfigRules>,
+    ) -> ZResult<PolicyInformation> {
+        let mut policy_rules: Vec<PolicyRule> = Vec::new();
+        for config_rule in config_rule_set {
+            for subject in &config_rule.interface {
+                for flow in &config_rule.flow {
+                    for action in &config_rule.action {
+                        for key_expr in &config_rule.key_expr {
+                            policy_rules.push(PolicyRule {
+                                subject: Subject::Interface(subject.clone()),
+                                key_expr: key_expr.clone(),
+                                action: *action,
+                                permission: config_rule.permission,
+                                flow: *flow,
+                            })
+                        }
+                    }
+                }
+            }
+        }
+        let mut subject_map = SubjectMap::default();
+        let mut counter = 1;
+        //starting at 1 since 0 is the init value and should not match anything
+        for rule in policy_rules.iter() {
+            subject_map.insert(rule.subject.clone(), counter);
```

Comment:
I also don’t really understand why this indirection was introduced for the interface; there don’t seem to be many rules and interfaces, but the additional indirection complicates the code.
Or is this some kind of foundation for the future?

Reply:
This is for performance in the future case where we might have a larger set of IDs in our rule-set.